### PR TITLE
Add tests

### PR DESCRIPTION
**Pull request by amichuda**
check X and y for dataframe and Series; check shape of X and shape of R; warn about full_enumeration

### CODE REVIEW

File: tests/test_bootalgo.py [hunk 1]
```diff
@@ -201,10 +215,46 @@ def test_error_warnings():
...
+      boot.get_denom()
+      boot.get_tboot()
+      boot.get_vcov()
+      boot.get_tstat()
+      boot.get_pvalue(pval_type = "two-tailed")
+      results_dict.append(boot.pvalue)
+        
+  results_series = pd.Series(results_dict)
+  mapd = results_series.mad() / results_series.mean()
+  assert  mapd <= .1 # make sure mean absolute percentage deviation is less than 10% (ad hoc)
```
Comment (s3alfisc) [inline tests/test_bootalgo.py:255]: I think we could set a lower threshold here, something in the 1-3% range? Also, this test will likely run for a while, but it's definitely good to have it.

**Review by s3alfisc — APPROVED** [review]
Looks good, feel free to merge!

File: tests/test_bootalgo.py (hunk 1 above)
Comment (amichuda) [reply tests/test_bootalgo.py:255]: We're actually getting that the percent change is 0.08. All the p-values are .12-.18, so inference is the same, but they're on average 8% apart. I cobbled together this metric as:

mean(|x-x.mean()|)/x.mean()

I'm not sure if it's a standard one, just something I made up. Am happy to change it to a different distance function and threshold though

**Comment by amichuda** [comment]
> Looks good, feel free to merge!

Yes, it looks like if this all works, then we have all the unit tests that we defined in #21 . After this all we need is CI, codecov, and we should publish this up to pypi

**Comment by s3alfisc** [comment]
Is it fine to publish on pypy while the tests for the WCU11, WCR1x still fail? I am still not sure what's going on there & will spend some time debugging on Saturday. Overall I think chances are quite low that I won't find the error if staring at the code for a few hours, so maybe we could set a soft goal to release on Sunday?

File: tests/test_bootalgo.py (hunk 1 above)
Comment (s3alfisc) [reply tests/test_bootalgo.py:255]: I'll take a closer look tomorrow / on Saturday - it's getting somewhat late for thorough review here in Germany =)

**Comment by amichuda** [comment]
Yeah, sure that works for me!

**Comment by amichuda** [comment]
Tests for the stochastic:

```
|                                             |   p_val_python |   p_val_R |
|:--------------------------------------------|---------------:|----------:|
| ('11', True, 'rademacher', 'two-tailed')    |      0.25      |    0.25   |
| ('11', True, 'rademacher', 'equal-tailed')  |      0.25      |    0.25   |
| ('11', True, 'rademacher', '>')             |      0.125     |    0.25   |
| ('11', True, 'rademacher', '<')             |      0.875     |    0.25   |
| ('11', True, 'mammen', 'two-tailed')        |      0.398894  |    0.25   |
| ('11', True, 'mammen', 'equal-tailed')      |      0.0428404 |    0.25   |
| ('11', True, 'mammen', '>')                 |      0.0216802 |    0.25   |
| ('11', True, 'mammen', '<')                 |      0.97848   |    0.25   |
| ('11', True, 'webb', 'two-tailed')          |      0.214322  |    0.25   |
| ('11', True, 'webb', 'equal-tailed')        |      0.211502  |    0.25   |
| ('11', True, 'webb', '>')                   |      0.105531  |    0.25   |
| ('11', True, 'webb', '<')                   |      0.892779  |    0.25   |
| ('11', True, 'norm', 'two-tailed')          |      0.190942  |    0.25   |
| ('11', True, 'norm', 'equal-tailed')        |      0.190902  |    0.25   |
| ('11', True, 'norm', '>')                   |      0.0933009 |    0.25   |
| ('11', True, 'norm', '<')                   |      0.903799  |    0.25   |
| ('11', False, 'rademacher', 'two-tailed')   |      0.1875    |    0.1875 |
| ('11', False, 'rademacher', 'equal-tailed') |      0.1875    |    0.1875 |
| ('11', False, 'rademacher', '>')            |      0.09375   |    0.1875 |
| ('11', False, 'rademacher', '<')            |      0.90625   |    0.1875 |
| ('11', False, 'mammen', 'two-tailed')       |      0.120021  |    0.1875 |
| ('11', False, 'mammen', 'equal-tailed')     |      0.138001  |    0.1875 |
| ('11', False, 'mammen', '>')                |      0.0672107 |    0.1875 |
| ('11', False, 'mammen', '<')                |      0.931019  |    0.1875 |
| ('11', False, 'webb', 'two-tailed')         |      0.201362  |    0.1875 |
| ('11', False, 'webb', 'equal-tailed')       |      0.202162  |    0.1875 |
| ('11', False, 'webb', '>')                  |      0.102791  |    0.1875 |
| ('11', False, 'webb', '<')                  |      0.896789  |    0.1875 |
| ('11', False, 'norm', 'two-tailed')         |      0.200942  |    0.1875 |
| ('11', False, 'norm', 'equal-tailed')       |      0.198362  |    0.1875 |
| ('11', False, 'norm', '>')                  |      0.101421  |    0.1875 |
| ('11', False, 'norm', '<')                  |      0.899809  |    0.1875 |
| ('31', True, 'rademacher', 'two-tailed')    |      0.25      |    0.25   |
| ('31', True, 'rademacher', 'equal-tailed')  |      0.25      |    0.25   |
| ('31', True, 'rademacher', '>')             |      0.125     |    0.25   |
| ('31', True, 'rademacher', '<')             |      0.875     |    0.25   |
| ('31', True, 'mammen', 'two-tailed')        |      0.399044  |    0.25   |
| ('31', True, 'mammen', 'equal-tailed')      |      0.0412404 |    0.25   |
| ('31', True, 'mammen', '>')                 |      0.0211302 |    0.25   |
| ('31', True, 'mammen', '<')                 |      0.97922   |    0.25   |
| ('31', True, 'webb', 'two-tailed')          |      0.212822  |    0.25   |
| ('31', True, 'webb', 'equal-tailed')        |      0.210742  |    0.25   |
| ('31', True, 'webb', '>')                   |      0.107361  |    0.25   |
| ('31', True, 'webb', '<')                   |      0.893339  |    0.25   |
| ('31', True, 'norm', 'two-tailed')          |      0.190632  |    0.25   |
| ('31', True, 'norm', 'equal-tailed')        |      0.190842  |    0.25   |
| ('31', True, 'norm', '>')                   |      0.0932509 |    0.25   |
| ('31', True, 'norm', '<')                   |      0.903949  |    0.25   |
| ('31', False, 'rademacher', 'two-tailed')   |      0.1875    |    0.1875 |
| ('31', False, 'rademacher', 'equal-tailed') |      0.1875    |    0.1875 |
| ('31', False, 'rademacher', '>')            |      0.09375   |    0.1875 |
| ('31', False, 'rademacher', '<')            |      0.90625   |    0.1875 |
| ('31', False, 'mammen', 'two-tailed')       |      0.120491  |    0.1875 |
| ('31', False, 'mammen', 'equal-tailed')     |      0.138701  |    0.1875 |
| ('31', False, 'mammen', '>')                |      0.0696807 |    0.1875 |
| ('31', False, 'mammen', '<')                |      0.931199  |    0.1875 |
| ('31', False, 'webb', 'two-tailed')         |      0.203642  |    0.1875 |
| ('31', False, 'webb', 'equal-tailed')       |      0.204302  |    0.1875 |
| ('31', False, 'webb', '>')                  |      0.101631  |    0.1875 |
| ('31', False, 'webb', '<')                  |      0.898319  |    0.1875 |
| ('31', False, 'norm', 'two-tailed')         |      0.198522  |    0.1875 |
| ('31', False, 'norm', 'equal-tailed')       |      0.199602  |    0.1875 |
| ('31', False, 'norm', '>')                  |      0.101101  |    0.1875 |
| ('31', False, 'norm', '<')                  |      0.899779  |    0.1875 |
```

**Comment by s3alfisc** [comment]
Urgh, this does not look as it should. Let's hope the tests are incorrect and not the algo :) one thing that is visible is that the pvalues produced by fwildclusterboot seem to be the same across different weights - maybe they are all accidentally slipping into the enumeration case? E.g. with only 5 clusters, there can a only 2^5 t stats, and 2^4 pvalues?

**Comment by s3alfisc** [comment]
I'll take a closer look today after work.

**Comment by s3alfisc** [comment]
The stochastic tests look better after some minor fixes to the tests: 

```

R p_val | Py  p_val
-- | --
0.123781 | 0.122781
0.124581 | 0.125621
0.063411 | 0.062511
0.937869 | 0.937949
0.127961 | 0.128521
... | ...
0.937979 | 0.938869
0.122791 | 0.124741
0.120801 | 0.122441
0.061801 | 0.061091
0.939399 | 0.936909
```

**Pull request merged**